Add Optional LOD Transition Period

nithinp7 opened the pull request:
Adds options to `TilesetOptions` to control a transition period between tiles of different LOD (when a tile is refined / unrefined). Currently a tile's transition percentage is available  in `TileContentLoadResult`, it can be used for client-implemented blending effects.

The Cesium for Unreal client implementation of dithered LOD transitions that leverages the changes in this PR is here: https://github.com/CesiumGS/cesium-unreal/pull/882

TODOs before merge:
* [x] Prohibit this feature from being used when culling is enabled? Previously this PR correctly handled tiles that are becoming un-culled. However, there is no good way to handle actual [holes](#269), so I'm considering only allowing the use of the feature when culling is disabled or forbid holes is enabled. 
* [x] Integrate these changes with Bao's refactor.

nithinp7 commented:
@kring As discussed offline, currently this feature overrides the "Enable Frustum / Fog Culling" options to false. This constraint is currently based on the assumption that culling will likely be off by default in the near future (to avoid [holes](#269)). However, we have also discussed implementing a mode where culling is enabled for quick loading, but auto-disabled once the visible tree is first finished loading (the first time the medium loading priority queue is empty?). 

So anyways, it may be worth expanding support for culling, I'm not sure. I had a previous version that worked well with tiles that were becoming "un-culled", but [holes](#269) still posed some issues. The best that could happen in the case of holes, is to pop them in at full opacity as soon as they are ready, which might be okay. Anyways, let me know if you think supporting load-culling would be worthwhile!

nithinp7 commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -1447,8 +1548,11 @@ Tileset::TraversalDetails Tileset::_visitTile(
   // The traversalDetails tell us what happened while visiting the children.
-  if (!traversalDetails.allAreRenderable &&
-      !traversalDetails.anyWereRenderedLastFrame) {
+  if ((!traversalDetails.allAreRenderable &&
+       !traversalDetails.anyWereRenderedLastFrame) ||
+      (lastFrameSelectionResult == TileSelectionState::Result::Rendered &&
+       tile.getContent() &&
+       tile.getContent()->lodTransitionFadePercentage < 1.0f)) {
```
@kring This is another part of the code worth reviewing in particular: the changes to the kick logic. This allows us to keep descendants off the render list if the tile is still fading in. The descendants may still continue to load though.

nithinp7 commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -1202,8 +1314,14 @@ Tileset::TraversalDetails Tileset::_visitTile(
   // The traversalDetails tell us what happened while visiting the children.
-  if (!traversalDetails.allAreRenderable &&
-      !traversalDetails.anyWereRenderedLastFrame) {
+  if ((!traversalDetails.allAreRenderable &&
+       !traversalDetails.anyWereRenderedLastFrame) ||
+      (_options.enableLodTransitionPeriod &&
+       _options.kickDescendantsWhileFadingIn &&
+       lastFrameSelectionResult == TileSelectionState::Result::Rendered &&
+       tile.getContent().isRenderContent() &&
+       tile.getContent().getRenderContent()->getLodTransitionFadePercentage() <
+           1.0f)) {
```
This part of the code is worth reviewing in particular: the changes to the kick logic. This allows us to keep descendants off the render list if the tile is still fading in. The descendants may still continue to load though. If this option is disabled, the fading-in tile will pop in at full opacity so that the rendered descendant can start fading in.

nithinp7 commented:
@baothientran This is integrated with the new changes in main and is ready to review when you get a chance.

baothientran reviewed (commented):
@nithinp7 I tried this out and the latency for the transition from coarse LOD to detailed LOD seems to be quite longer than I would like but that maybe about to adjust the settings. Another thing is that when I tried this with PIE mode in unreal, I can see both the fade in and fade out tiles fully opaque like below, so I'm not sure if it's expected:

![Translation](https://user-images.githubusercontent.com/20376598/186275023-fd8baf03-6e35-4772-8bfe-14aa4f0a9a4d.gif)


Since this option turns off both frustum and fog culling, it maybe worth to measure the number of requests and memory between on and off. The draw calls maybe the same since unreal does some frustum culling behind the scene as well

nithinp7 commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -235,6 +243,108 @@ Tileset::updateView(const std::vector<ViewState>& frustums) {
+    for (Tile* pTile : result.tilesToNoLongerRenderThisFrame) {
+      if (pTile->getContent().isRenderContent()) {
+        pTile->getContent().getRenderContent()->setLodTransitionFadePercentage(
+            2.0f);
```
I don't know if it is hacky, but basically fading-in tiles count up from 0 to 1, and fading out tiles count down from 2 to 0. This lets the fading-out tile wait to fade until the fading-in tile is fully opaque, without the two tiles actually knowing about each other.

nithinp7 commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -165,10 +165,11 @@ Tileset::updateViewOffline(const std::vector<ViewState>& frustums) {
 
-  this->updateView(frustums);
+  // TODO: fix the fading for offline case
+  this->updateView(frustums, 0.0f);
```
Opened one here: https://github.com/CesiumGS/cesium-native/issues/549

nithinp7 commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -235,6 +243,108 @@ Tileset::updateView(const std::vector<ViewState>& frustums) {
+    }
+
+    float deltaTransitionPercentage =
+        2.0f * deltaTime / this->_options.lodTransitionLength;
```
This is for the same reason as I answered in the other comment:
>I don't know if it is hacky, but basically fading-in tiles count up from 0 to 1, and fading out tiles count down from 2 to 0. This lets the fading-out tile wait to fade until the fading-in tile is fully opaque, without the two tiles actually knowing about each other.

baothientran replied:
Can you please add this comment into the code to explain why the number 2, so that we don't forget it?

nithinp7 replied:
There's this comment a few lines above it:
>// We fade the old tile from 2.0 --> 0.0 while the new tile fades from
    // 0.0 --> 1.0. The old tile doesn't appear faded until it is < 1.0.

nithinp7 commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/ViewUpdateResult.h:
```diff
@@ -31,6 +35,10 @@ class CESIUM3DTILESSELECTION_API ViewUpdateResult final {
 
+  std::unordered_set<Tile*> tilesFadingOut;
+
+  std::vector<Tile*> tilesToHideThisFrame;
```
I changed the semantics / naming of the containers in `ViewUpdateResult` and also added documentation. I think all four containers are necessary, but can you check out the new documentation that describes the usage of each one? Let me know if you still think it could be simplified.

baothientran replied:
Not sure if this works but can we simplify to only two fields `tilesFadeInThisFrame` and `tilesFadeOutThisFrame`? `tilesFadeInThisFrame` requires the client to render the tile with alpha opacity. And `tilesFadeOutThisFrame` requires the client to render the tile with alpha opacity > 0 or hide the tile when alpha opacity < 0 and also disable the collision for fade-out tiles.

nithinp7 replied:
Good idea, I applied this suggestion in the latest commit! Now the fading in tiles + the fully opaque tiles are in the `tilesToRenderThisFrame`. The fading out tiles are in `tilesFadingOut` (renaming of `tilesToNoLongerRender`). The client is responsible for hiding tiles when their `lodTransitionPercentage` is 0 - the next frame these tiles won't appear on the list, so the client needs to hide them as soon as they reach 0. 

In the case of Cesium for Unreal, the physics for tiles on the `tileFadingOut` are disabled the first time they appear on the list. Later, if they appear on the render list again, their physics meshes are reenabled.

baothientran commented:
@nithinp7 do you know if occlusion culling will work if alpha mask is on in the shader? Can you check the performance of that

nithinp7 commented:
@baothientran Our default materials already use opacity mask for clipping. The parts of the material graph that contribute to the opacity mask also get used during the depth-only prepass. In this case, the dithering causes one extra texture sample (I assume the dithering node is sampling some sort of noise texture). Other than that, I don't think occlusion culling's effectiveness or performance is affected much. Note that dithering doesn't use translucency, it's just an opacity mask effect - translucency in an object disqualifies it from being drawn as an occluder, that would have a much bigger affect on occlusion culling performance / effectiveness.

nithinp7 commented:
@baothientran In the last commit I simplified the `ViewUpdateResult` struct based on your comments (left a more detailed comment about this in the conversation above). I also moved the weird [0-2] fading range logic to Cesium for Unreal, since it is specific to dithering which is a client implementation detail. Now the transition progress percentage is very simple, fading out tiles go from 1 --> 0 and fading in tiles go from 0 --> 1. The client side can remap this range as needed.

I looked a bit at performance impact, I'm not positive but I think there may be a temporary increase in draw calls in some scenes when moving the camera around. This is an inevitable cost to LOD blending, but in Cesium for Unreal I am looking into if we can stop the fading-in tiles from contributing to the shadow pass, which seem to be the heaviest impact.

I think disabling culling definitely increases load latency for the target detail to show up, but I don't think it makes too much of an impact on memory usage. We have "preload siblings" on by default, which converges to the same amount of memory usage once the low priority tiles are all loaded. The only difference is the prioritization and the latency of high-detail tiles getting loaded. 

Yet another rant about holes:
So the tradeoff as I see it is latency vs holes, and in the context of fade transitions the holes are really egregious sometimes. The fading-out tile stays opaque until the fading-in tile _should_ be loaded (remember they don't actually know anything about each other). Once that point happens, the tile starts fading out. But if there's a hole the fading-in tile may still not be there! In some cases, the effect is that you turn around quickly and there are no holes initially but then suddenly a tile starts fading out to reveal a hole after you've stopped moving. This brings even more attention to holes than if we had just turned around and a hole just flashed on the screen. The only fix (other than disabling culling) would be to implement a system where tile refinements / unrefinements are explicitly tracked and transitions can be aborted if a hole appears on the screen, which would create a popping around the hole that would emphasize the hole even more!

nithinp7 commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -1288,6 +1406,22 @@ void Tileset::_unloadCachedTiles() noexcept {
+        pTile->getContent()
+                .getRenderContent()
+                ->getLodTransitionFadePercentage() > 0.0f) {
+      pTile = this->_loadedTiles.next(*pTile);
```
I think the difference in memory usage due to the transition period shouldn't be much of an issue unless the tileset is already pushing up against the in-memory cache limit in a certain view. Regarding the memory impact caused by disabling culling (quoted from a comment below):

>I think disabling culling definitely increases load latency for the target detail to show up, but I don't think it makes too much of an impact on memory usage. We have "preload siblings" on by default, which converges to the same amount of memory usage once the low priority tiles are all loaded. The only difference is the prioritization and the latency of high-detail tiles getting loaded.

kring commented on Cesium3DTilesSelection/test/TestTilesetSelectionAlgorithm.cpp:
```diff
@@ -876,8 +874,7 @@ TEST_CASE("Test multiple frustums") {
       REQUIRE(result.tilesToRenderThisFrame.size() == 4);
 
-      REQUIRE(result.tilesToNoLongerRenderThisFrame.size() == 1);
-      REQUIRE(result.tilesToNoLongerRenderThisFrame.front() == root);
```
Why was the second check removed?

kring reviewed (commented):
Looks good to me, just a few minor comments. Actually they're so minor I'll probably just make the changes myself after I submit this, and then merge the PR.
